[#1]; feat: 회원가입 기능 추가

Sennyru opened the pull request:
<!-- 이슈번호를 작성해주세요 -->
Closes #1 

간단한 회원가입 API를 설계했습니다.

## 구현 내용
`/api/members/signup` URI에 아래와 같은 POST 요청을 보내면 DB에 회원 데이터가 저장되고 로그인 정보를 받습니다.

[예시 요청]
```json
{
    "email" : "email@urssu.com",
    "password" : "password",
    "username" : "username"
}
```
[예시 응답]
```json
{
    "email" : "email@urssu.com",
    "username" : "username"
}
```


만약 이미 존재하는 이메일로 회원가입을 시도할 경우 아래와 같은 응답을 받습니다.
```json
{
  "status": "BAD_REQUEST",
  "message": "이미 가입된 이메일입니다.",
  "requestURI": "/api/members/signup",
  "time": "2025-09-25T02:24:26.0886478"
}
```

비밀번호는 Spring Security 라이브러리로 암호화하여 저장합니다.

## 종속성
- lombok, JPA, Spring Security, H2
- Spring Boot Validation은 아직 사용하지 않지만 미리 추가

Sennyru committed:
feat: 회원가입 API 구현

- domain, DTO, repository, service, controller 등 기본적인 API 계층 설계
- lombok, JPA, Spring Security(비밀번호 암호화 용), H2 종속성 추가
- Spring Boot Validation은 아직 사용하지 않지만 미리 추가
- 테스트 필요

Sennyru committed:
fix: API를 누구나 접근 가능하게 변경

- Spring Security 라이브러리로 인해, 기본적으로 로그인을 해야 API 요청을 보낼 수 있었음
- 이를 로그인이 필요하지 않도록 보안 규칙 수정

Sennyru committed:
fix: 중복 이메일 생성 방지

- 사용자 DB에는 중복 이메일이 존재할 수 없음
- DB에 존재하는 이메일과 같은 이메일로 요청이 들어왔을 경우, 기존에는 응답이 오지 않았음
- 이제 400 Bad Request 응답

seulnan commented on src/main/java/com/sennyru/onboarding/controller/MemberController.java:
```diff
@@ -0,0 +1,28 @@
+    private final MemberService memberService;
+
+    @PostMapping("/signup")
+    public ResponseEntity<MemberResponseDto> signup(@RequestBody SignupRequestDto requestDto) {
```
suggestion: 요기에서 제안한 @Valid 어노테이션 한번 사용해보는것이 좋을것같아요.

Sennyru replied:
`@Valid`를 이용한 입력 유효성 검사 코드를 추가했습니다! ([해당 커밋](https://github.com/Sennyru/Backend-Study-Onboarding/commit/a9b058e88ac223c0f5f3f4fd757bc0234fe6aa8c))

seulnan commented on src/main/java/com/sennyru/onboarding/domain/Member.java:
```diff
@@ -0,0 +1,38 @@
+
+@Entity
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
이 어노테이션은 어떻게 작동되나요?? 그리고 왜 써야하나요??

Sennyru replied:
* `@Entity` - 해당 클래스를 엔티티(DB의 테이블에 해당)로 설정합니다.
* `@Getter` - 각 필드의 getter를 자동으로 생성해 줍니다(lombok 라이브러리). 일일이 getter를 적는 것보다 훨씬 깔끔하고 타이핑이 빠르기 때문에 사용합니다.
* `@NoArgsConstructor(...)` - 클래스의 기본 생성자를 protected로 생성(=변경)합니다. 기본 생성자를 외부에서 사용하지 못하게 하는 역할을 합니다.

seulnan commented on src/main/java/com/sennyru/onboarding/service/MemberService.java:
```diff
@@ -0,0 +1,35 @@
+
+        String encryptedPassword = passwordEncoder.encode(requestDto.getPassword());
+
+        Member member = Member.builder()
+                .email(requestDto.getEmail())
+                .password(encryptedPassword)
+                .username(requestDto.getUsername())
+                .build();
```
빌더패턴을 사용하신 이유가 따로 있을까요? 저는 객체생성의 의도를 더 명확하게 표현하기 위해 정적팩토리메서드를 사용하는걸 선호하긴합니다.

Sennyru replied:
특별한 이유는 없습니다. 제안대로 빌더 대신 정적 팩토리 메서드로 바꾸었습니다! ([해당 커밋](https://github.com/Sennyru/Backend-Study-Onboarding/commit/39d4eb43f110dcfd1070af1b90688b2351b0ed08))

seulnan commented on src/main/java/com/sennyru/onboarding/service/MemberService.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Transactional
+    public Member signup(SignupRequestDto requestDto) {
+        if (memberRepository.existsByEmail(requestDto.getEmail())) {
+            throw new IllegalStateException("이미 가입된 이메일입니다.");
+        }
```
코파일럿이 얘기해줬듯이 레이스 컨디션이 발생할 수도 있네요. 왜 발생할 수도 있을까요? 레이스 컨디션이 뭘까요? 그리고 어떻게 해결해야할까요?

Sennyru replied:
레이스 컨디션은 여러 프로세스 또는 스레드가 공유 자원을 사용할 때 발생하는 문제입니다. 이 코드에서는 스레드 A에서 중복 이메일 체크를 하고 DB에 계정을 저장하기 직전에 스레드 B가 끼어드는 경우, 스레드 B에서 중복 이메일이 없다 판단하고 두 스레드가 각자 DB에 계정을 저장하려 시도하게 됩니다.
이 프로젝트의 경우 email 필드가 UNIQUE하기 때문에 큰 문제가 발생하지는 않으나, 예외 처리가 되지 않아 HTTP Response가 제대로 되지 않습니다. 그래서 DB에서 UNIQUE 원칙을 위반했을 때 발생하는 예외를 적절히 처리해서 ErrorResponse를 던져 주면 해결됩니다.
다른 상황에서 레이스 컨디션이 발생하는 경우, 메서드를 synchronized로 바꾸거나, DB에 락을 걸고 트랜잭션이 끝나면 락을 해제하거나, DB의 버전을 체크하는 방식으로 해결할 수 있습니다. (노션에 조금 더 자세히 적어놨습니다)

seulnan reviewed (changes requested):
너무 고생하셨습니다 !
질문이 많은데, 잘 공부했는지 확인하기 위함이니, 공부하신대로 답해주시면 됩니다~~

Sennyru committed:
[#1]; refactor: Member & MemberResponseDto 객체 생성에 정적 팩토리 메서드 패턴 적용

- 기존에 두 클래스의 인스턴스를 생성하려면 생성자를 이용해야 했음.
- 그러나 이 방식은 객체 생성의 책임이 각 클래스 바깥에 있는 문제가 있음.
- 그래서 정적 팩토리 메서드 패턴을 도입하여 객체 생성에 대한 책임을 클래스 자신에게 두고 객체 생성의 의도를 더욱 명확하게 만듦.

Sennyru commented on src/main/java/com/sennyru/onboarding/controller/MemberController.java:
```diff
@@ -0,0 +1,28 @@
+    @PostMapping("/signup")
+    public ResponseEntity<MemberResponseDto> signup(@RequestBody SignupRequestDto requestDto) {
+        Member savedMember = memberService.signup(requestDto);
+        MemberResponseDto responseDto = new MemberResponseDto(savedMember);
```
정적 팩토리 메서드를 도입하여 객체 생성의 책임을 DTO 자신으로 돌렸습니다! ([해당 커밋](https://github.com/Sennyru/Backend-Study-Onboarding/commit/39d4eb43f110dcfd1070af1b90688b2351b0ed08))

Sennyru committed:
[#1]; refactor: SignupRequestDto & ErrorResponse 객체 생성에도 정적 팩토리 메서드 패턴 적용

- 이전 커밋에서는 Member와 MemberResponseDto에만 정적 팩토리 메서드를 적용했었음.
- SignupRequestDto와 ErrorResponse에도 코드 일관성을 위해 정적 팩토리 메서드를 적용함.

Sennyru committed:
[#1]; fix: signup 메서드에서 레이스 컨디션 발생 시 예외 처리
- 원인: MemberService의 signup 메서드에서 레이스 컨디션이 발생할 여지가 있음.
- DB의 email 속성이 UNIQUE하기 때문에 레이스 컨디션이 발생해도 문제가 발생하지는 않지만, 예외 처리가 되지 않음.
- 해결: UNIQUE 위반 시 발생하는 DataIntegrityViolationException 예외를 처리하여 CONFLICT 코드를 응답하게 함.

Sennyru commented on src/main/java/com/sennyru/onboarding/service/MemberService.java:
```diff
@@ -0,0 +1,35 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional(readOnly = true)
```
**왜?**
트랜잭션은 과정이 하나라도 실패하면 모든 과정을 원래대로 되돌립니다. 이를 통해 작업의 일관성을 지키기 위해서 트랜잭션을 사용합니다.

**어떻게?**
DB를 업데이트하는 메서드에 `@Transactional` 어노테이션을 붙여서 사용합니다. 읽기만 하는 작업의 경우 `readonly = true` 옵션을 붙이면 성능이 향상됩니다.

**안티패턴?**
- `@Transactional`은 `RuntimeException`이 발생했을 때에만 롤백이 됩니다. 그래서 오류 상황에 대한 예외 처리를 잘못하면 의도와는 다르게 롤백이 되지 않을 수 있습니다.
- 트랜잭션 코드의 범위를 너무 넓게 잡으면 DB를 너무 오랫동안 점유하여 프로그램 성능이 저하될 수 있습니다.

Sennyru committed:
[#1]; fix: `@Valid`를 사용하여 API 입력 유효성 검사
- 현재 상황: /signup 요청에 대한 입력 값의 유효성 검사를 하지 않음.
- 수정: 컨트롤러에 `@Valid` 어노테이션을 추가하여 입력 필드의 형식이 제대로 됐는지 검사함.
- 공백 필드 방지, 이메일 형식 체크

Sennyru committed:
[#1]; refactor: DTO를 record 클래스로 변경
- 클래스 코드를 깔끔하게 하기 위하여 DTO들을 record 클래스로 변경함

Sennyru commented on src/main/java/com/sennyru/onboarding/dto/MemberResponseDto.java:
DTO들을 record 클래스로 변경했습니다. ([해당 커밋](https://github.com/Sennyru/Backend-Study-Onboarding/commit/130f5b7436019bef07d438ab8b99efb7160198e1))

Sennyru committed:
[#1]; refactor: DTO-엔티티 간 의존성 분리
- 기존 코드의 경우 MemberResponseDto가 직접 Member 매개변수를 받아서 DTO 객체를 생성했었음.
- 이러한 DTO와 엔티티 간의 의존 관계를 분리함.
- 엔티티를 DTO로 바꾸는 코드는 Service 계층으로 이전.

Sennyru commented on src/main/java/com/sennyru/onboarding/dto/MemberResponseDto.java:
```diff
@@ -0,0 +1,15 @@
+    private String email;
+    private String username;
+
+    public MemberResponseDto(Member member) {
```
엔티티가 변경됐을 때 DTO가 영향을 받는 문제가 생깁니다. DTO는 API 명세를 담당하는 중요한 부분이므로 영향을 최소화하는 것이 좋습니다.
그래서 MemberResponseDto 객체를 생성할 때 Member 매개변수를 사용하지 않도록 변경했습니다. ([해당 커밋](https://github.com/Sennyru/Backend-Study-Onboarding/commit/7429f50741f682bdfbcec5bd1d9f143f9b1f80a4)) 이제 엔티티를 DTO로 변경하는 일은 Service가 하게 됩니다.
